### DISCUSSION TITLE
I'm not sure that I understand but can zydis output assembly/object files?

### DISCUSSION THREAD

**Discussion by rempas** [zyantific/zydis#355] [Q&A]
The description says:
> Fast and lightweight x86/x86-64 disassembler and code generation library

But what "code generation" exactly means? Does this mean that it generates either assembly or (even better) machine code? It may be obvious but I didn't see an IR and the examples only show the "encoding" and "decoding" stuff.

**Comment by mappzor** [comment]
"Code generation" means "encoding" i.e. generating machine code.

**Reply by flobernd** [reply]
@rempas To elaborate a little further: Yes, you can output object files, PE files, and so on, but that requires some manual work from your side. Zydis is very low level and only emits instructions.

If you are looking for a more high level assembler, you can check out [zasm](https://github.com/zyantific/zasm). Output for PE files is being under development.

**Reply by rempas** [reply]
Thank you for the info! In the [Encoder Example](https://github.com/zyantific/zydis#encoder-example) section in the README.md, it shows an example but the output is hex numbers. This is what I don't understand, cat it also output a file directly?

If not, then how do I turn it into an executable? Should I create the ELF (for Unix) files manually and the place the hex code manually?

**Reply by rempas** [reply]
@flobernd Thank you! I can wait for PE files as I'm currently interested in ELF files (and my frontend is very early in development anyway). Another question that raises from this is that, If I'm using 2 libraries, what will be the compile time performance? Do you yet have an benchmarks in this regard?

**Reply by flobernd** [reply]
> Should I create the ELF (for Unix) files manually and the place the hex code manually?

Yes. The hex output in the encoder sample is just a human readable representation of the machine code.

> Another question that raises from this is that, If I'm using 2 libraries, what will be the compile time performance? Do you yet have an benchmarks in this regard?

Zasm is build on top of Zydis, so it will effectively be the same library 😋 Regarding exact numbers and ELF support, probably @ZehMatt has some answers for you.

**Reply by rempas** [reply]
Thanks for everything! I find myself enjoying creating a frontened more and more but the backend is another thing.... Not only it's hard to learn the instructions but at the same time, you have different instructions for different platforms, you must make the optimizations, etc. And keep in mind that all that must be in a professional level as I don't want to make a toy language.

So yeah, making all these will take me years so finding a library and using the work and effort some has already make is a win-win!

**Reply by ZehMatt** [reply]
Sorry that I'm late to the party, I have not yet published the "frontend" part what I call it, I just finished the PE builder part recently demonstrated here https://gist.github.com/ZehMatt/626eb4e7f78bbff74d2615308bdf7d87
The code currently relies on the LIEF library and has no support for ELF from scratch but with zasm you have enough information to build most common binary/object/linker files if you really want to.

**Reply by rempas** [reply]
@ZehMatt Thanks for the info my friend!
